modules/services/views/src/main/java/au/com/shiftyjelly/pocketcasts/views/multiselect/MultiSelectToolbar.kt

```diff
@@ -30,42 +30,54 @@ class MultiSelectToolbar @JvmOverloads constructor(
 ) : Toolbar(context, attrs, defStyleAttr) {
 
     companion object {
+        private const val MAX_ICONS_BOOKMARKS = 2
         const val MAX_ICONS = 4
     }
 
     private var overflowItems: List<MultiSelectAction> = emptyList()
-    private var fragmentManager: FragmentManager? = null
-    private var multiSelectHelper: MultiSelectHelper? = null
     @Inject lateinit var analyticsTracker: AnalyticsTrackerWrapper
 
-    fun setup(
+    fun <T> setup(
         lifecycleOwner: LifecycleOwner,
-        multiSelectHelper: MultiSelectHelper,
+        multiSelectHelper: MultiSelectHelper<T>,
         @MenuRes menuRes: Int?,
         fragmentManager: FragmentManager
     ) {
-
-        this.fragmentManager = fragmentManager
-        this.multiSelectHelper = multiSelectHelper
-
         setBackgroundColor(context.getThemeColor(UR.attr.support_01))
         if (menuRes != null) {
             inflateMenu(menuRes)
         } else {
             multiSelectHelper.toolbarActions.observe(lifecycleOwner) {
                 menu.clear()
 
-                it.subList(0, MAX_ICONS).forEachIndexed { _, action ->
+                val maxIcons = when (multiSelectHelper) {
+                    is MultiSelectBookmarksHelper -> MAX_ICONS_BOOKMARKS
+                    is MultiSelectEpisodesHelper -> MAX_ICONS
+                    else -> 0
+                }
+                it.subList(0, maxIcons).forEachIndexed { _, action ->
                     val item = menu.add(Menu.NONE, action.actionId, 0, action.title)
                     item.setIcon(action.iconRes)
                     item.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+                    item.isVisible = action.isVisible
                 }
 
-                overflowItems = it.subList(MAX_ICONS, it.size)
-
-                val overflow = menu.add(Menu.NONE, R.id.menu_overflow, 0, context.getString(LR.string.more_options))
-                overflow.setIcon(IR.drawable.ic_more_vert_black_24dp)
-                overflow.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+                overflowItems = it.subList(maxIcons, it.size)
+
+                when (multiSelectHelper) {
+                    is MultiSelectBookmarksHelper -> {
+                        overflowItems.forEachIndexed { _, action ->
+                            val item = menu.add(Menu.NONE, action.actionId, 0, action.title)
+                            item.setIcon(action.iconRes)
+                            item.setShowAsAction(MenuItem.SHOW_AS_ACTION_NEVER)
+                        }
+                    }
+                    is MultiSelectEpisodesHelper -> {
+                        val overflow = menu.add(Menu.NONE, R.id.menu_overflow, 0, context.getString(LR.string.more_options))
+                        overflow.setIcon(IR.drawable.ic_more_vert_black_24dp)
+                        overflow.setShowAsAction(MenuItem.SHOW_AS_ACTION_ALWAYS)
+                    }
```

Comment:
Why is there different logic for bookmarks and episodes? I'm not quite following this part.

Reply:
It's a single menu item, and doesn't need any edit action, that's why.